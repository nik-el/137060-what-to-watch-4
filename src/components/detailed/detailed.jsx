import React from 'react';
import PropTypes from 'prop-types';
import {FilmPropTypes} from "../../types/film-prop-types";
import {useParams} from 'react-router-dom';

export const Detailed = React.memo(function Detailed({films}) {
  const {id} = useParams();

  console.log('idididid:', id);
  const film = films.find((item) => item.id === id);
  const {
    title,
    genre,
    releaseYear,
    thumbnail,
    poster,
    description,
    rating,
    ratingCount,
    directors,
    starring
  } = film;

  return <section className="movie-card movie-card--full">
    <div className="movie-card__hero">
      <div className="movie-card__bg">
        <img src={thumbnail} alt="The Grand Budapest Hotel"/>
      </div>

      <div className="movie-card__wrap">
        <div className="movie-card__desc">
          <h2 className="movie-card__title">{title}</h2>
          <p className="movie-card__meta">
            <span className="movie-card__genre">{genre}</span>
            <span className="movie-card__year">{releaseYear}</span>
          </p>

          <div className="movie-card__buttons">
            <button className="btn btn--play movie-card__button" type="button">
              <svg viewBox="0 0 19 19" width="19" height="19">
                <use xlinkHref="#play-s"/>
              </svg>
              <span>Play</span>
            </button>
            <button className="btn btn--list movie-card__button" type="button">
              <svg viewBox="0 0 19 20" width="19" height="20">
                <use xlinkHref="#add"/>
              </svg>
              <span>My list</span>
            </button>
            <a href="add-review.html" className="btn movie-card__button">Add review</a>
          </div>
        </div>
      </div>
    </div>

    <div className="movie-card__wrap movie-card__translate-top">
      <div className="movie-card__info">
        <div className="movie-card__poster movie-card__poster--big">
          <img
            src={poster}
            alt={title} width="218"
            height="327"/>
        </div>

        <div className="movie-card__desc">
          <nav className="movie-nav movie-card__nav">
            <ul className="movie-nav__list">
              <li className="movie-nav__item movie-nav__item--active">
                <a href="#" className="movie-nav__link">Overview</a>
              </li>
              <li className="movie-nav__item">
                <a href="#" className="movie-nav__link">Details</a>
              </li>
              <li className="movie-nav__item">
                <a href="#" className="movie-nav__link">Reviews</a>
              </li>
            </ul>
          </nav>

          <div className="movie-rating">
            <div className="movie-rating__score">{rating}</div>
            <p className="movie-rating__meta">
              <span className="movie-rating__level">Very good</span>
              <span className="movie-rating__count">{ratingCount} ratings</span>
            </p>
          </div>

          <div className="movie-card__text">
            <p>{description}</p>
            <p className="movie-card__director"><strong>Director: {directors[0]}</strong></p>
            <p className="movie-card__starring"><strong>Starring: {starring[0]}</strong></p>
          </div>
        </div>
      </div>
    </div>
  </section>;
});

Detailed.propTypes = {
  films: PropTypes.arrayOf(PropTypes.shape(FilmPropTypes)),
};
