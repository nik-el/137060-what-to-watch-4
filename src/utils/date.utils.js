const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const getFormattedDate = (dateString) => {
  if (!dateString) {
    return ``;
  }

  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
