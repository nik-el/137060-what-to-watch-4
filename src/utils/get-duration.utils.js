const HOUR_MIN = 60;
const HOUR_SEC = 3600;

export const geFormatDuration = (time) => {
  const hours = Math.floor(time / HOUR_MIN);
  const min = time % HOUR_MIN;

  return `${hours}h ${min}m`;
};

export const geVideoDuration = (time) => {
  const secNum = parseInt(time, 10);
  let hours = Math.floor(secNum / HOUR_SEC);
  let minutes = Math.floor((secNum - (hours * HOUR_SEC)) / HOUR_MIN);
  let seconds = secNum - (hours * HOUR_SEC) - (minutes * HOUR_MIN);

  if (hours < 10) {
    hours = `0` + hours;
  }
  if (minutes < 10) {
    minutes = `0` + minutes;
  }
  if (seconds < 10) {
    seconds = `0` + seconds;
  }
  return hours + `:` + minutes + `:` + seconds;
};
