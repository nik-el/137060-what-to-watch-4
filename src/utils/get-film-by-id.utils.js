export const getFilmById = (films, id) => {
  if (!Array.isArray(films) || !id) {
    return null;
  }

  return films.find((item) => item.id === id);
};
