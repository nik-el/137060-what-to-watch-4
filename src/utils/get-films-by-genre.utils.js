export const getFilmsByGenre = (films, genre) => {
  if (!Array.isArray(films)) {
    return [];
  }
  if (!genre) {
    return films;
  }

  return films.filter((item) => item.genre === genre);
};
