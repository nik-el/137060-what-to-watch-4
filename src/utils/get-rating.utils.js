export const getRating = (rating) => {
  switch (true) {
    case (rating < 3):
      return `Bad`;
    case (rating >= 3 && rating < 5):
      return `Normal`;
    case (rating >= 5 && rating < 8):
      return `Good`;
    case (rating >= 8 && rating < 10):
      return `Very good`;
    case (rating === 10):
      return `Awesome`;
    default:
      return `No rating`;
  }
};
