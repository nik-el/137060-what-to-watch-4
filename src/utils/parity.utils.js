const getOddArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }
  return array.filter((item, index) => {
    return index % 2;
  });

};
const getEvenArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }

  return array.filter((item, index) => {
    return !(index % 2);
  });
};

export {
  getOddArray,
  getEvenArray
};
