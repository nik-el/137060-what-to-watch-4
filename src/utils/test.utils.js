const testFilm = {
  title: `Test title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testFilm);

const testEditedFilm = {
  title: `Test edit title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testEditedFilm);

const testFilms = [
  {
    title: `Test title 1`,
    genre: `test genre 1`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/1`,
    bgImage: `https://picsum.photos/id/1`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/1`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `1`,
    preview: `https://picsum.photos/id/1`,
    runTime: 100,
  },
  {
    title: `Test title 2`,
    genre: `test genre 2`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/2`,
    bgImage: `https://picsum.photos/id/2`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `2`,
    preview: `https://picsum.photos/id/2`,
    runTime: 100,
  },
];
Object.freeze(testFilms);

const testGenres = [`test genre 1`, `test genre 2`];
Object.freeze(testGenres);

const testComments = [{
  id: 1,
  user: {
    id: 4,
    name: `Test  name`
  },
  rating: 1.1,
  comment: `test comment`,
  date: `2019-05-08T14:13:56.569Z`
}];
Object.freeze(testComments);

const testStore = {
  DATA: {
    films: testFilms,
    genres: testGenres,
    loadingFilmsData: false,
    restApi: `test_api_url`
  },
  VIEW: {
    currentGenre: testGenres[0],
    feedLimit: 10,
  },
  USER: {
    authorizationStatus: `TEST_AUTH`,
    user: {
      "id": 1,
      "email": `test@test.com`,
      "name": `test`,
      "avatarUrl": `test/1.png`
    }
  },
  REVIEW: {
    loadingCommentsData: false,
    loadingAddingComment: false,
    comments: testComments
  },
  FAVORITE: {
    favorites: testFilms,
    loadingFavoritesData: false,
    loadingSetFavorite: false
  }
};
Object.freeze(testStore);

export {testFilm, testFilms, testEditedFilm, testGenres, testComments, testStore};
