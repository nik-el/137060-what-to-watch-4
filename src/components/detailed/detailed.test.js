import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter, Route} from "react-router-dom";
import {Detailed} from './detailed';

import {testFilms} from '../../utils/test.utils';
const currentTestFilmId = testFilms[0].id;

it(`Detailed renders correctly`, () => {
  const tree = renderer
    .create(
        <MemoryRouter initialEntries={[`detailed/${currentTestFilmId}`]}>
          <Route path={`detailed/:id`}>
            <Detailed
              films={testFilms}
              match={{params: {id: currentTestFilmId}}}
            />
          </Route>
        </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
