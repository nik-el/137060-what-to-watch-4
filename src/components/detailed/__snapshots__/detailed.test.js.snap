// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detailed renders correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__hero"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="https://picsum.photos/id/1"
      />
    </div>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__desc"
      >
        <h2
          className="movie-card__title"
        >
          Test title 1
        </h2>
        <p
          className="movie-card__meta"
        >
          <span
            className="movie-card__genre"
          >
            Comedy
          </span>
          <span
            className="movie-card__year"
          />
        </p>
        <div
          className="movie-card__buttons"
        >
          <button
            className="btn btn--play movie-card__button"
            type="button"
          >
            <svg
              height="19"
              viewBox="0 0 19 19"
              width="19"
            >
              <use
                xlinkHref="#play-s"
              />
            </svg>
            <span>
              Play
            </span>
          </button>
          <button
            className="btn btn--list movie-card__button"
            type="button"
          >
            <svg
              height="20"
              viewBox="0 0 19 20"
              width="19"
            >
              <use
                xlinkHref="#add"
              />
            </svg>
            <span>
              My list
            </span>
          </button>
          <a
            className="btn movie-card__button"
            href="add-review.html"
          >
            Add review
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="movie-card__wrap movie-card__translate-top"
  >
    <div
      className="movie-card__info"
    >
      <div
        className="movie-card__poster movie-card__poster--big"
      >
        <img
          alt="Test title 1"
          height="327"
          src="https://picsum.photos/id/2"
          width="218"
        />
      </div>
      <div
        className="movie-card__desc"
      >
        <nav
          className="movie-nav movie-card__nav"
        >
          <ul
            className="movie-nav__list"
          >
            <li
              className="movie-nav__item movie-nav__item--active"
            >
              <a
                className="movie-nav__link"
                href="#"
              >
                Overview
              </a>
            </li>
            <li
              className="movie-nav__item"
            >
              <a
                className="movie-nav__link"
                href="#"
              >
                Details
              </a>
            </li>
            <li
              className="movie-nav__item"
            >
              <a
                className="movie-nav__link"
                href="#"
              >
                Reviews
              </a>
            </li>
          </ul>
        </nav>
        <div
          className="movie-rating"
        >
          <div
            className="movie-rating__score"
          >
            5,5
          </div>
          <p
            className="movie-rating__meta"
          >
            <span
              className="movie-rating__level"
            >
              Very good
            </span>
            <span
              className="movie-rating__count"
            >
              100
               ratings
            </span>
          </p>
        </div>
        <div
          className="movie-card__text"
        >
          <p>
            Test description
          </p>
          <p
            className="movie-card__director"
          >
            <strong>
              Director: 
              Test director
            </strong>
          </p>
          <p
            className="movie-card__starring"
          >
            <strong>
              Starring: 
              Test Star 1
            </strong>
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;
